use fs-extra copy instead of ncp

develar opened the pull request:
Closes #213, #150

electron-builder tests passed as well.

malept commented on docs/api.md:
```diff
@@ -220,7 +220,7 @@ Object (also known as a "hash") of application metadata to embed into the execut
 ### `err`
 
-*Error* (or *Array*, in the case of an `ncp` error)
+*Error* (or *Array*, in the case of an `copy` error)
```
Does `copy` actually create an `Array` on erroring? Because that would be very unfortunate.

malept replied:
I answered my own question below.

develar replied:
We can pass `stopOnError: true` to stop on first error.

malept replied:
We should probably update all of the `copy` calls to use that (in a separate PR). Having `err` be potentially more than one type is problematic for users.

malept commented on package.json:
```diff
@@ -36,10 +35,10 @@
     "rcinfo": "^0.1.3",
     "run-waterfall": "^1.1.1",
-    "standard": "^5.4.1",
+    "standard": "^6.0.8",
```
Do you mind removing these dependency updates from this PR?

develar replied:
Strictly speaking, it is an unrelated change, so, sorry, I will remove and address in a separate PR.

develar replied:
reverted.

malept commented on test/fixtures/basic/package.json:
```diff
@@ -5,7 +5,6 @@
     "run-series": "^1.1.1"
   },
   "devDependencies": {
-    "ncp": "^2.0.0",
```
I added this particular dependency to test #186. If you remove it, please replace it with a module (with zero dependencies) that creates a file in `node_modules/.bin`.

develar replied:
reverted and comment added.

malept commented:
@develar I looked into how `fs-extra` implemented `copy`, because you didn't change any of the options passed. It turns out that it's [just a wrapper for `ncp`](https://github.com/jprichardson/node-fs-extra/blob/d863f5fa00db751ff5faa73ae8a15344da677d29/lib/copy/copy.js#L3). Could you explain how this change would fix all the issues you mention?

malept commented:
Hm. I just realized that it uses a forked copy of ncp. But just because it may have some fixes in it, doesn't necessarily mean that all of those issues are resolved. We also don't have any tests for #150 or #213. If you want to assert that those are resolved, please write testcases for them.

malept commented on test/fixtures/basic/package.json:
```diff
@@ -4,6 +4,8 @@
     "run-series": "^1.1.1"
   },
+  "//": "ncp used to test https://github.com/electron-userland/electron-packager/pull/186",
+  "///": "(a module (with zero dependencies) that creates a file in node_modules/.bin)",
```
That is a really unfortunate way to annotate JSON files.

develar replied:
Yes, but npm doesn't support other ways (e.g. typescript supports comments in the tsconfig.json) :(

malept commented:
By the way, it might be better to squash your commits when the PR has been approved. It makes it easier for reviewers to see what changed, especially now that GitHub has added features to the pull request UI.

develar commented:
There is a comment https://github.com/AvianFlu/ncp/issues/98#issuecomment-173273082 so, I think, it is definitely fixed in the fs-extra. I don't see any particular file sets in the #150 and #213 so, I am not sure how I can test it. May be we can trust @jprichardson and existing fs-extra [tests](https://github.com/jprichardson/node-fs-extra/tree/master/lib/copy/__tests__)?

develar commented:
> it might be better to squash your commits

Ok, I will not amend commits anymore.

malept commented:
I'm not so sure the ncp issue you mention is fixed in fs-extra. If it is, it wasn't obvious from the commit log. My reading of his comment was more along the lines of "here's a maintained version of ncp".

jprichardson commented:
> I'm not so sure the ncp issue you mention is fixed in fs-extra. If it is, it wasn't obvious from the commit log. My reading of his comment was more along the lines of "here's a maintained version of ncp".

To my knowledge, I've fixed most of the outstanding issues in my forked copy of `ncp`. If there's anything you're unsure of, let me know and I can take a second look.

malept commented:
@jprichardson the issue I'm unsure that it fixes is https://github.com/AvianFlu/ncp/issues/98. Another look through the history (GitHub does not make that easy due to the multiple renames) indicates that it may be the same as https://github.com/jprichardson/node-fs-extra/issues/98 (coincidentally enough)?

jprichardson commented:
https://github.com/AvianFlu/ncp/issues/98 leaves out a lot of details, so it's unclear of the problem. The commenter just states that the callback is never being called with no code to reproduce. Who knows why. I put a ton of work into fixing the issues in `ncp` (as seen in https://github.com/jprichardson/node-fs-extra/issues/98) and others.

Also, I can state that I use `electron-packager` for my startup (http://www.exodus.io - depends upon Electron), so I'm heavily invested in the Electron ecosystem (if it matters).

malept commented:
@jprichardson @develar OK, sounds good to me. At minimum, it gets rid of an extra dependency (I've already gotten rid of direct production dependencies on `mv`, `rimraf`, and `mkdirp` in favor of `fs-extra` since this PR was filed).

@develar once you rebase, LGTM.

malept commented:
@develar FYI your build failed, seems [you missed some `ncp` references](https://travis-ci.org/electron-userland/electron-packager/jobs/122167908#L1628-L1629).

jprichardson commented:
Oh, I wanted to add one additional point. If it matters, `fs-extra` will be dropping support for Node v0.10 within the next few months. I wouldn't think this would be a big deal, but if you're looking to continue with Node v0.10 support (I couldn't imagine why), you'll be stuck in time with the last version of `fs-extra` to support Node v0.10. I don't see how this situation is any worse than the unmaintained `ncp` though. I just wanted to make sure you had all info up front before you make your decision.

malept commented:
It's been a while since we've supported 0.10 officially. I think I had to remove support for it in Travis CI for some feature. At any rate, I filed a PR the other day that [definitely removes support for Node < 4.0](https://github.com/electron-userland/electron-packager/pull/319). So we're OK :smile:

develar commented:
Sorry, fixed.

maxogden commented:
Just wanted to point out this doesnt close https://github.com/electron-userland/electron-packager/issues/151, that issue is about using tar-fs, which IMO is a lot more dependable performance wise than fs-extra/ncp.

malept commented:
Removed #151 from the issue summary.

malept commented:
@develar squash and fix the commit message to omit #151 and I will merge.

jprichardson commented:
> which IMO is a lot more dependable performance wise than fs-extra/ncp.

Would love to know why so that I can improve `fs-extra`.

Overall though, `tar-fs`, `cpy`, `fs-extra`... puh-tay-toe/puh-tah-toe. I don't care much as long as we get a better `electron-packager` and I'm happy to assist.

develar commented:
> which IMO is a lot more dependable performance wise than fs-extra/ncp.

It is a bit illogical for me. How does tar work — read file and append content to resulting (total) file. And then we will read total file and split it to destination files.

So, why it will be faster? copy operation just transfer bytes from one file to another.

Or do you mean that copy is not reliable? Well... if nobody cannot write reliable copy implementation on node platform... But I don't think so — I have tried 3 different implementations and only fs-extra copy works for me (electron-builder tests — copying test fixtures). And recently windows-installer got rid of jetpack and switched to fs-extra.

So, #151 was mentioned not by mistake.

develar commented:
Apart of this, I am going to get rid of copy phase (copy app files) at all :) Because if `prune` is not specified, we can build asar archive directly, without intermediate copy.

malept commented:
That "squash and merge" feature in GitHub's UI is pretty great. Won't work all the time, but worked well in this instance.
